pygfx/renderers/wgpu/meshrender.py

```diff
@@ -113,22 +94,190 @@ def mesh_renderer(wobject, render_info):
                 func.__annotations__["t_tex"] = tex_anno
             fragment_shader = fragment_shader.float_version
 
-    if index_buffer:
-        n = len(index_buffer.data)
-    else:
-        n = len(vertex_buffers[0].data)
-
     # Put it together!
 
     return [
         {
             "vertex_shader": vertex_shader,
             "fragment_shader": fragment_shader,
-            "primitive_topology": wgpu.PrimitiveTopology.triangle_list,
+            "primitive_topology": topology,
             "indices": (range(n), range(1)),
             "index_buffer": index_buffer,
             "vertex_buffers": vertex_buffers,
             "bindings0": bindings0,
             "bindings1": bindings1,
         }
     ]
+
+
+# Taken from Vispy
+def _calculate_normals(rr, tris):
+    """Efficiently compute vertex normals for triangulated surface"""
+    # ensure highest precision for our summation/vectorization "trick"
+    rr = rr[:, :3].astype(np.float64)
+    tris = tris.reshape(-1, 3)
+    # first, compute triangle normals
+    r1 = rr[tris[:, 0], :]
+    r2 = rr[tris[:, 1], :]
+    r3 = rr[tris[:, 2], :]
+    tri_nn = np.cross((r2 - r1), (r3 - r1))
+
+    # Triangle normals and areas
+    size = np.sqrt(np.sum(tri_nn * tri_nn, axis=1))
+    size[size == 0] = 1.0  # prevent ugly divide-by-zero
+    tri_nn /= size[:, np.newaxis]
+
+    npts = len(rr)
+
+    # the following code replaces this, but is faster (vectorized):
+    #
+    # for p, verts in enumerate(tris):
+    #     nn[verts, :] += tri_nn[p, :]
+    #
+    nn = np.zeros((npts, 3))
+    for verts in tris.T:  # note this only loops 3x (number of verts per tri)
+        for idx in range(3):  # x, y, z
+            nn[:, idx] += np.bincount(
+                verts.astype(np.int32), tri_nn[:, idx], minlength=npts
+            )
```

Comment:
as an alternative, you could also write it like this if you keep a reference to the un-normalized `tri_nn`:

```
for i in range(3):
    np.add.at(nn, tris[:, i], tri_nn)
```

not sure if it's faster/equivalent/slower, but it's certainly a lot easier to read